working/2364 - primary constructors/feature-specification.md

```diff
@@ -361,14 +397,28 @@ constants *(holding parameter default values)*.
 Where no processing is mentioned below, _D2_ is identical to _D_. Changes
 occur as follows:
 
+Assume that `p` is an optional formal parameter in _D_ that does not have a
+declared type, but it does have a default value whose static type in the
+empty context is a type `T` which is not `Null`. In that case `p` is
+considered to have the declared type `T`. When `T` is `Null`, `p` is
+considered to have the declared type `Object?`.
+
 The current scope of the formal parameter list of the primary constructor
 in _D_ is the current scope of the class/enum declaration *(in other words,
 the default values cannot see declarations in the class body)*. Every
```

Comment:
This scope differs from in-body constructors, but I can see how it's *partly* consistent with the position. Might want to check if this is what users expect. Maybe moving an in-body field declaration `String x = _defaultString;` into the primary constructor parameter list should "just work", and not require rewriting as `String x = MyClass._defaultString`.

The expression *should* be inside the parameter scope, because that's where i syntactically is.
That means:
```dart
static const x = 0, T = 0, C = 0;
class C<T>(final int x, [int y = x + T + C]);
```
should reject all of `x`, `T` and `C` because they are in scope for the expression (they're declared before and around the expression, so they should be in the lexical scope for it).
That can still be outside the body scope of the class, but it's not the surrounding scope of the declaration either (which I think "current scope" means).

Reply:
Definitely. Done.